src/fragments/lib/graphqlapi/js/optimistic-ui.mdx

```diff
@@ -0,0 +1,539 @@
+
+<Callout>
+
+See the [Documentation](https://tanstack.com/query/latest/docs/react/guides/optimistic-updates) for more information on optimistic updates.
+
+</Callout>
+
+To create a mutation, use the `useMutation` hook, passing in the GraphQL mutation (and variables). The following example creates a mutation to create a new record via the API.
+
+```ts
+// Mutations
+const mutation = useMutation({
+  mutationFn: postTodo,
+  onSuccess: () => {
+    // Invalidate and refetch
+    queryClient.invalidateQueries({ queryKey: ['todos'] })
+  },
+})
+```
+
+## Update
+
+## Delete
+
+## Loading states, etc
+
+TODO:
+
+## Debugging optimistic UI
+
+**NOTE: this could also be a callout**
+
+To help visualize the inner workings of React Query, visit the [TanStack Query DevTools docs](https://tanstack.com/query/latest/docs/react/devtools) for more information.
+
+## Complete examples
+
+<BlockSwitcher>
+<Block name="TypeScript">
+
+```ts
+import React, { useState } from "react";
+import "./App.css";
+import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
+import { API } from "aws-amplify";
+import * as queries from "./graphql/queries";
+import * as mutations from "./graphql/mutations";
+import { GraphQLQuery } from "@aws-amplify/api";
+import {
+  CreateRealEstatePropertyInput,
+  CreateRealEstatePropertyMutation,
+  DeleteRealEstatePropertyInput,
+  DeleteRealEstatePropertyMutation,
+  GetRealEstatePropertyQuery,
+  ListRealEstatePropertiesQuery,
+  RealEstateProperty,
+  UpdateRealEstatePropertyInput,
+  UpdateRealEstatePropertyMutation,
+} from "./API";
+import { Loader } from "@aws-amplify/ui-react";
+
+// https://tanstack.com/query/v5/docs/react/guides/background-fetching-indicators#displaying-global-background-fetching-loading-state
+import { useIsFetching } from "@tanstack/react-query";
+
+// For rendering 
+function GlobalLoadingIndicator() {
+  const isFetching = useIsFetching();
+
+  return isFetching ? <Loader size="large" /> : null;
+}
+
+// For generating random address
+const alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+function App() {
+  const [currentRealEstatePropertyId, setCurrentRealEstatePropertyId] =
+    useState<string | null>(null);
+
+  // Access the client
+  const queryClient = useQueryClient();
+
+  // TanStack Query for listing all realEstateProperties
+  const {
+    data: realEstateProperties,
+    isLoading,
+    isSuccess,
+    isError: isErrorQuery,
+  } = useQuery({
+    queryKey: ["realEstateProperties"],
+    // TODO: improve return type
+    queryFn: async (): Promise<
+      (RealEstateProperty | null | undefined)[] | null | undefined
+    > => {
+      const response = await API.graphql<
+        GraphQLQuery<ListRealEstatePropertiesQuery>
+      >({
+        query: queries.listRealEstateProperties,
+      });
+
+      const allRealEstateProperties =
+        response?.data?.listRealEstateProperties?.items;
+
+      if (!allRealEstateProperties) return null;
```

Comment:
maybe `return []`

Reply:
I'll add a todo here - I'll sort out types once we settle on the final approach for the code example.